File: src/services/deviceUserInfo.ts
/**
 * Device/User Info Handler (ap2014)
 *
 * Operation-aware service for retrieving subscriber information.
 * Supports GetPhoneNumber and GetSubscriberInfo operations.
 */

import type { ApplicationConfig } from '../protocol/responseTypes.js';
import { SubscriptionResult } from '../protocol/statusCodes.js';
import type { OdsaContext } from './odsaCommon.js';
import { buildOdsaBaseConfig } from './odsaCommon.js';

export interface DeviceUserInfoConfigData {
  msisdn?: string;
  displayName?: string;
  homeCarrier?: string;
}

export function buildDeviceUserInfoConfig(
  status: number,
  provStatus: number,
  tcStatus: number,
  configData?: unknown,
  odsaContext?: OdsaContext,
): ApplicationConfig {
  const data = (configData ?? {}) as DeviceUserInfoConfigData;
  const operation = odsaContext?.operation;

  if (!operation) {
    return {
      appId: 'ap2014',
      entitlementStatus: status,
      provStatus,
      tcStatus,
    };
  }

  switch (operation) {
    case 'GetPhoneNumber':
      return handleGetPhoneNumber(status, provStatus, tcStatus, data);
    case 'GetSubscriberInfo':
      return handleGetSubscriberInfo(status, provStatus, tcStatus, data);
    default:
      return {
        appId: 'ap2014',
        entitlementStatus: status,
        provStatus,
        tcStatus,
      };
  }
}

function handleGetPhoneNumber(
  status: number, provStatus: number, tcStatus: number, data: DeviceUserInfoConfigData,
): ApplicationConfig {
  const extra: Record<string, string> = {};
  if (data.msisdn) extra.MSISDN = data.msisdn;
  return buildOdsaBaseConfig('ap2014', status, provStatus, tcStatus, SubscriptionResult.DONE, extra);
}

function handleGetSubscriberInfo(
  status: number, provStatus: number, tcStatus: number, data: DeviceUserInfoConfigData,
): ApplicationConfig {
  const extra: Record<string, string> = {};
  if (data.msisdn) extra.MSISDN = data.msisdn;
  if (data.displayName) extra.DisplayName = data.displayName;
  if (data.homeCarrier) extra.HomeCarrier = data.homeCarrier;
  return buildOdsaBaseConfig('ap2014', status, provStatus, tcStatus, SubscriptionResult.DONE, extra);
}

File: internal/services/device_user_info.go
package services

import (
	"github.com/hkassaei/entitlements-as-a-service-poc-go/internal/protocol"
)

// DeviceUserInfoConfigData holds configData for ap2014.
type DeviceUserInfoConfigData struct {
	MSISDN      string `json:"msisdn,omitempty"`
	DisplayName string `json:"displayName,omitempty"`
	HomeCarrier string `json:"homeCarrier,omitempty"`
}

// BuildDeviceUserInfoConfig builds the ApplicationConfig for Device/User Info (ap2014).
func BuildDeviceUserInfoConfig(status, provStatus, tcStatus int, configData *DeviceUserInfoConfigData, ctx *OdsaContext) *protocol.ApplicationConfig {
	if configData == nil {
		configData = &DeviceUserInfoConfigData{}
	}

	operation := ""
	if ctx != nil {
		operation = ctx.Operation
	}

	if operation == "" {
		return &protocol.ApplicationConfig{
			AppID:             "ap2014",
			EntitlementStatus: status,
			ProvStatus:        protocol.IntPtr(provStatus),
			TcStatus:          protocol.IntPtr(tcStatus),
		}
	}

	switch operation {
	case "GetPhoneNumber":
		extra := map[string]string{}
		if configData.MSISDN != "" {
			extra["MSISDN"] = configData.MSISDN
		}
		return BuildOdsaBaseConfig("ap2014", status, provStatus, tcStatus, protocol.SubscriptionResultDone, extra)
	case "GetSubscriberInfo":
		extra := map[string]string{}
		if configData.MSISDN != "" {
			extra["MSISDN"] = configData.MSISDN
		}
		if configData.DisplayName != "" {
			extra["DisplayName"] = configData.DisplayName
		}
		if configData.HomeCarrier != "" {
			extra["HomeCarrier"] = configData.HomeCarrier
		}
		return BuildOdsaBaseConfig("ap2014", status, provStatus, tcStatus, protocol.SubscriptionResultDone, extra)
	default:
		return &protocol.ApplicationConfig{
			AppID:             "ap2014",
			EntitlementStatus: status,
			ProvStatus:        protocol.IntPtr(provStatus),
			TcStatus:          protocol.IntPtr(tcStatus),
		}
	}
}
